File: ar.lua
--------------------------------------------------------------------------------
-- ATTACK-RELEASE ENVELOPE
--------------------------------------------------------------------------------

-- knobs
ar_attack = 0
ar_release = 0

-- internal
ar_state = 0
ar_done = 0
ar_todo = 0

-- out
ar_out = 0

-- derived from update_adsr()
function update_ar()
    if kbd_trigger then
        ar_state = 1
        ar_done = 0
        ar_todo = math.floor(ar_attack * sr)
    end
    if ar_state == 1 then
        if ar_todo > 0 then
            ar_out = ar_done / (ar_done + ar_todo)
        else
            ar_state = 3
        end
    end
    if ar_state == 3 then
        ar_out = 1
    end
    if not kbd_gate and ar_state ~= 0 and ar_state ~= 4 then
        ar_state = 4
        ar_done = 0
        ar_todo = math.floor(ar_release * sr)
    end
    if ar_state == 4 then
        if ar_todo > 0 then
            ar_out = ar_todo / (ar_done + ar_todo)
        else
            ar_state = 0
        end
    end
    if ar_state == 0 then
        ar_out = 0
    else
        ar_done = ar_done + 1
        ar_todo = ar_todo - 1
    end
end

File: ar.cpp
////////////////////////////////////////////////////////////////////////////////
// ATTACK-RELEASE ENVELOPE
////////////////////////////////////////////////////////////////////////////////

#include <cmath>

#include "synth.hpp"

// derived from update_adsr()
void Synth::update_ar() {
    auto auto_trigger = (adsr.kbd_repeat and kbd.gate and lfo.trigger) or (not adsr.kbd_repeat and lfo.trigger);
    auto trigger = (ar.kbd_trigger and kbd.trigger) or (not ar.kbd_trigger and auto_trigger);
    if (trigger) {
        ar.state = 1;
        ar.done = 0;
        ar.todo = std::floor(ar.attack * sr);
    }
    if (ar.state == 1) {
        if (ar.todo > 0) {
            ar.out = ar.done / (ar.done + ar.todo);
        } else {
            ar.state = 3;
        }
    }
    if (ar.state == 3) {
        ar.out = 1;
    }
    if ((ar.kbd_trigger and not kbd.gate and ar.state != 0 and ar.state != 4) or (not ar.kbd_trigger and ar.state == 3)) {
        ar.state = 4;
        ar.done = 0;
        ar.todo = std::floor(ar.release * sr);
    }
    if (ar.state == 4) {
        if (ar.todo > 0) {
            ar.out = ar.todo / (ar.done + ar.todo);
        } else {
            ar.state = 0;
        }
    }
    if (ar.state == 0) {
        ar.out = 0;
    } else {
        ar.done = ar.done + 1;
        ar.todo = ar.todo - 1;
    }
}
